src/Storages/MergeTree/MergeTreeIndexText.cpp

```diff
@@ -1336,37 +1415,70 @@ Tuple parseNamedArgumentFromAST(const ASTFunction * ast_equal_function)
         result.emplace_back(identifier->name());
     }
 
-    /// Parse parameter value. It can be Literal, Identifier or Function.
-    if (const auto * literal_arg = arguments->children[1]->as<ASTLiteral>(); literal_arg != nullptr)
-    {
-        result.emplace_back(literal_arg->value);
-    }
-    else if (const auto * identifier_arg = arguments->children[1]->as<ASTIdentifier>(); identifier_arg != nullptr)
+    if (result.back() == "preprocessor")
     {
-        result.emplace_back(identifier_arg->name());
+        const ASTFunction * preprocessor_function = arguments->children[1]->as<ASTFunction>();
+        if (preprocessor_function == nullptr)
+            throw Exception(ErrorCodes::INCORRECT_QUERY, "Text index preprocessor argument must be an expression");
+
+        String identifier_name;
+        validatePreprocessorASTExpression(preprocessor_function, identifier_name);
```

Comment:
can we make this part generic but not `preprocessor` specific? This validation is kinda part of semantic analysis not the parsing. I'd prefer to keep parsing and semantic analysis steps separately.

Reply:
Yes, that is the preprocessor semantic analysis but my point is that semantic analysis should be after the parsing process fully completed. Here we are trying to parse and do semantic analysis at the same time. The initial phase of parsing happens from `IndicesDescription.cpp:112` which calls `validatePreprocessorASTExpression` transitively during this phase. Then in `textIndexValidator` we call `validatePreprocessorASTExpression` again (from `MergeTreeIndexTextPreprocessor::parseExpression`).
IMO the first call from `IndicesDescription.cpp:112` should be as simple as possible (just collect identifiers, literals, tuples, etc) and number of arguments can vary.

All other validations should happen at `textIndexValidator`.

BUT it's up to you. This is also fine.